alibi_detect/cd/base_online.py

```diff
@@ -8,24 +8,98 @@
 import numpy as np
 from alibi_detect.base import BaseDetector, concept_drift_dict
 from alibi_detect.cd.utils import get_input_shape
-from alibi_detect.utils.frameworks import Framework, has_pytorch, has_tensorflow
+from alibi_detect.utils.frameworks import Framework
 from alibi_detect.utils._state import save_state_dict, load_state_dict
 from alibi_detect.utils._types import Literal
 
-if has_pytorch:
-    import torch
-
-if has_tensorflow:
-    import tensorflow as tf
-
 logger = logging.getLogger(__name__)
 
 
-class BaseMultiDriftOnline(BaseDetector):
+class BaseDriftOnline(BaseDetector):
+    """
+    Base class for online drift detectors.
+    """
+    t: int = 0
     thresholds: np.ndarray
-    backend: Literal['pytorch', 'tensorflow']
     online_state_keys: Tuple[str, ...]
+    backend: Optional[Literal['pytorch', 'tensorflow']] = None
+
+    @abstractmethod
+    def _configure_thresholds(self):
+        pass
+
+    @abstractmethod
+    def _update_state(self, x_t):
+        pass
```

Comment:
Type hints of parameters and return types required.

Reply:
This is unfortunately necessary, since univariate online detectors have `def _update_state(self, x_t: np.ndarray):`  whilst multivariate have `def _update_state(self, x_t: torch.Tensor):` (or `tf.Tensor`). We violate Liskov's substitution principle slightly.

I sort of think this is OK to not add type hints in the abstract method since we only have it there to signal that sub-classes must have an `_update_method` which takes an instance and updates online state, but we don't specify the exact type. However, we could also do `def _update_state(self, x_t: Union[np.ndarray, 'torch.Tensor', 'tf.Tensor'):` and then add `# type: ignore[override]` in the sub-class? This is actually what we did previously...

P.s. its a similar story for the `get_threshold` method...

Reply:
@jklaise @mauicv I have got rid of the new `BaseDriftOnline` class. There are stil some existing issues like the ones above, but this PR now shouldn't introduce new ones at least! 

The `save_state`/`load_state` methods are now added via `StateMixin`.